ckmame.man
.\" Converted with mdoc2man 0.2
.\" from NiH: ckmame.mdoc,v 1.13 2005/06/12 19:12:26 wiz Exp 
.\" $NiH: ckmame.mdoc,v 1.13 2005/06/12 19:12:26 wiz Exp $
.\"
.\" Copyright (c) 2003, 2004, 2005 Dieter Baron and Thomas Klausner.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above
.\"    copyright notice, this list of conditions and the following
.\"    disclaimer in the documentation and/or other materials provided
.\"    with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THOMAS KLAUSNER ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH CKMAME 1 "June 12, 2005" NiH
.SH "NAME"
ckmame \- verify MAME rom sets
.SH "SYNOPSIS"
.B ckmame
[\fB-bcdFfhKkLlnSsUuVvw\fR]
[\fB-D\fR \fBdbfile\fR]
[\fBgame ...\fR]
.SH "DESCRIPTION"
.B ckmame
is a program to check ROM sets for MAME, the Multiple Arcade
Machine Emulator (see
\fBhttp://www.mame.net/ )\fR,
or other emulators that provide ROM set information in a compatible
way (e.g., Raine, Impact, or other emulators.)
It tells you which ROM files are missing or have a wrong checksum,
can delete unknown and unused files from the ROM sets, and rename
or move files.
.PP
MAME is an emulator for numerous arcade machines (video games, ranging
from the 1970s to today).
Since MAME is emulating the games, it uses the original images from
the arcade machines' original ROM chips.  MAME's internal database
contains extensive information on the ROMs, including the sizes
and checksums of the used roms.
Since some arcade games are variants of others and share some of the
same ROMs, there are also internal dependencies between these so
called clones.
To save space, ROMs can be archived in zip files and
dynamically unzipped by MAME when you play a game.
.B ckmame
does only support ROM sets stored as zip files.
.PP
.B ckmame
tries to create a
``split ROM set''
from your current collection.
This means that all files needed by the parent are kept in the
parent's zip file, and only the files needed by the clone (or child)
that are different from the parent's are kept in the child's zip file.
For identifying the ROMs ckmame uses the sizes and checksums; names are
only used for verification, since the MAME team often renames ROMs.
.PP
While
.B ckmame
uses MAME's rom database, it must be converted into a more
efficient format before use.
See
mkmamedb(1)
for details.
.PP
By default,
.B ckmame
will check the games named on the command line, or all games if none
are given.
Shell wildcards are allowed.
You can use command line options to select which type of errors to report.
.SS "OPTIONS"
Supported options:
.RS
.TP 24
Fl b, \-nobroken
Don't report unfixable errors.
See also
\fB-f\fR
and
\fB-w\fR.
.TP 24
\fB-c\fR, \-correct
Report correct sets.
Normally they are not mentioned.
.TP 24
\fB-D\fR \fBdbfile, \fB--db\fR Ar dbfile\fR
Use
\fBdbfile\fR
as database file.
.TP 24
\fB-d\fR, \-nonogooddumps
Don't report roms with no known good dumps.
.TP 24
\fB-F\fR, \-fix
Fix rom sets; do as much as possible, such as renaming and moving
files.
.TP 24
\fB-f\fR, \-nofixable
Don't report fixable errors.
See also
\fB-b\fR
and
\fB-w\fR.
.TP 24
\fB-h\fR, \-help
Display a short help message.
.TP 24
\fB-i\fR, \-integrity
Check integrity of zip archives and disk images
by computing checksums and comparing them with the
internally saved ones.
.TP 24
\fB-K\fR, \-keep-unknown
Keep unknown files when fixing (default).
They are moved into a zip file with the same name under
\fBroms/garbage\fR.
See also
\fB-k\fR.
.TP 24
\fB-k\fR, \-delete-unknown
Don't keep unknown files when fixing.
See also
\fB-K\fR.
.TP 24
\fB-L\fR, \-keep-long
Keep too long files when fixing (default).
See also
\fB-K\fR
and
\fB-l\fR.
.TP 24
\fB-l\fR, \-delete-long
Don't keep long files when fixing, delete them instead.
See also
\fB-L\fR.
.TP 24
\fB-n\fR, \-dryrun
Don't actually fix anything, just report what would be done.
.TP 24
\fB-S\fR, \-samples
Check samples instead of roms.
.TP 24
\fB--superfluous\fR
Only check for superfluous files in rom sets.
.TP 24
\fB-s\fR, \-nosuperfluous
Don't report superfluous (unneeded) files.
.TP 24
\fB-U\fR, \-keep-unused
Keep unused files when fixing.
Unused files are files that should, and do exist in other places.
See also
\fB-u\fR.
.TP 24
\fB-u\fR, \-delete-unused
Don't keep unused files when fixing (default), delete them instead.
See also
\fB-U\fR.
.TP 24
\fB-V\fR, \-version
Display program name and version number.
.TP 24
\fB-v\fR, \-verbose
Print fixes made.
.TP 24
\fB-w\fR, \-nowarnings
Print only unfixable error, like missing or completely incorrect roms.
See also
\fB-b\fR
and
\fB-f\fR.
.TP 24
\fB-X\fR, \-ignoreextra
Do not mention extra files that are in the roms or samples dirs
respectively.
Most of the time these are romsets whose names have changed.
Additionally,
.B ckmame
creates a
\fBroms/garbage\fR
directory (see also
\fB-K\fR ),
which you will also be warned about.
.RE
.SH "ENVIRONMENT"
.RS
.TP 12
Ev MAMEDB
Location for the database file.
Can be overridden with the
\fB-D\fR
option.
Defaults to
``\fBmame.db\fR.''
.TP 12
Ev ROMPATH
Location for the rom zip files.
Defaults to
``\fB.\fR.''
.RE
.SH "EXAMPLES"
Print a report about the current state of your rom sets in the
\fBroms\fR
subdir of the current dir:
.IP
\fBckmame\fR
.PP
.PP
Fix all sets and only report unfixable errors, omitting those
roms where no good dump exists:
.IP
\fBckmame \-Fwd\fR
.PP
.SH "SEE ALSO"
dumpgame(1),
mkmamedb(1),
xmame(6)
.SH "AUTHORS"

.B ckmame
was written by
Dieter Baron
<dillo@giga.or.at>
and
Thomas Klausner
<tk@giga.or.at.>
.SH "BUGS"
Files that are in clone sets of a set, but would belong in the parent
set, are not moved automatically.
Add them to the parent set manually, then run
\fBckmame \-F\fR
to remove them from the clone set(s) automatically.


NEWS
0.6 (2005-06-17)

- libzip is distributed separately
- Support MAME/MESS -listxml output
- Adapt for MAME -listinfo output changes (honor flags baddump and nodump).
- Check disk images (chd files).
- Superfluous merged into ckmame. ckmame now by default reports
  extra files.  Also works for samples.
- add file format version to database and don't use incompatible database
- dumpgame: find game by checksum of one of its ROMs
- add script to create database by running xmame

0.5 (2003-01-30)

MESS, Impact, Raine, and others based on the CMPro format now supported.
Man page added.
Bug fixes.

0.4.1

If a a clone zip didn't exist, it was not created.
The DOS port now really can fix romsets.

0.4

The whole program.


ckmame.c
/*
  $NiH: ckmame.c,v 1.39 2005/06/12 18:00:59 wiz Exp $

  ckmame.c -- main routine for ckmame
  Copyright (C) 1999, 2003, 2004, 2005 Dieter Baron and Thomas Klausner

  This file is part of ckmame, a program to check rom sets for MAME.
  The authors can be contacted at <nih@giga.or.at>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License, version 2, as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/



#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <fnmatch.h>

#include "config.h"

#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#else
#include "getopt.h"
#endif

#include "types.h"
#include "dbh.h"
#include "util.h"
#include "funcs.h"
#include "error.h"



char *prg;

char *usage = "Usage: %s [-hVSwsfbdcFKkUuLlvn] [-D dbfile] [game...]\n";

char help_head[] = PACKAGE " by Dieter Baron and Thomas Klausner\n\n";

char help[] = "\n"
"  -b, --nobroken       don't report unfixable errors\n"
"  -c, --correct        report correct sets\n"
"  -D, --db dbfile      use mame-db dbfile\n"
"  -d, --nonogooddumps  don't report roms with no good dumps\n"
"  -F, --fix            fix rom sets\n"
"  -f, --nofixable      don't report fixable errors\n"
"  -h, --help           display this help message\n"
"  -i, --integrity      check integrity of rom files and disk images\n"
"  -K, --keep-unknown   keep unknown files when fixing (default)\n"
"  -k, --delete-unknown don't keep unknown files when fixing\n"
"  -L, --keep-long      keep long files when fixing (default)\n"
"  -l, --delete-long    don't keep long files when fixing\n"
"  -n, --dryrun         don't actually fix, only report what would be done\n"
"  -S, --samples        check samples instead of roms\n"
"      --superfuous     only check for superfluous files in rom sets\n"
"  -s, --nosuperfluous  don't report superfluous files in rom sets\n"
"  -U, --keep-unused    keep unused files when fixing\n"
"  -u, --delete-unused  don't keep unused files when fixing (default)\n"
"  -V, --version        display version number\n"
"  -v, --verbose        print fixes made\n"
"  -w, --nowarnings     print only unfixable errors\n"
"  -X, --ignoreextra    ignore extra files in rom/samples dirs\n"
"\nReport bugs to <nih@giga.or.at>.\n";

char version_string[] = PACKAGE " " VERSION "\n"
"Copyright (C) 2003, 2004 Dieter Baron and Thomas Klausner\n"
PACKAGE " comes with ABSOLUTELY NO WARRANTY, to the extent permitted by law.\n"
"You may redistribute copies of\n"
PACKAGE " under the terms of the GNU General Public License.\n"
"For more information about these matters, see the files named COPYING.\n";

#define OPTIONS "bcD:dFfhiKkLlnSsxUuVvwX"

#define OPT_SF	256

struct option options[] = {
    { "help",          0, 0, 'h' },
    { "version",       0, 0, 'V' },

    { "correct",       0, 0, 'c' }, /* +CORRECT */
    { "db",            1, 0, 'D' },
    { "delete-long",   0, 0, 'l' },
    { "delete-unknown",0, 0, 'k' },
    { "delete-unused" ,0, 0, 'u' },
    { "dryrun",        0, 0, 'n' },
    { "fix",           0, 0, 'F' },
    { "ignoreextra",   0, 0, 'X' },
    { "integrity",     0, 0, 'i' },
    { "keep-long",     0, 0, 'L' },
    { "keep-unknown",  0, 0, 'K' },
    { "keep-unused",   0, 0, 'U' },
    { "nobroken",      0, 0, 'b' }, /* -BROKEN */
    { "nofixable",     0, 0, 'f' }, /* -FIX */
    { "nonogooddumps", 0, 0, 'd' }, /* -NO_GOOD_DUMPS */
    { "nosuperfluous", 0, 0, 's' }, /* -SUP */
    { "nowarnings",    0, 0, 'w' }, /* -SUP, -FIX */
    { "samples",       0, 0, 'S' },
    { "superfluous",   0, 0, OPT_SF },
    { "verbose",       0, 0, 'v' },

    { NULL,            0, 0, 0 },
};

int output_options;
int fix_do, fix_print, fix_keep_long, fix_keep_unused, fix_keep_unknown;
int ignore_extra;
int romhashtypes, diskhashtypes;



int
main(int argc, char **argv)
{
    int i, j;
    DB *db;
    char **list, *dbname;
    int c, nlist, found, dbext;
    struct tree *tree;
    struct tree tree_root;
    int sample, superfluous_only, integrity;
    
    prg = argv[0];
    tree = &tree_root;
    tree->child = NULL;
    output_options = WARN_ALL;
    sample = 0;
    superfluous_only = 0;
    dbext = 0;
    dbname = getenv("MAMEDB");
    if (dbname == NULL) {
	dbname = "mame";
	dbext = DDB_EXT;
    }
    fix_do = fix_print = 0;
    fix_keep_long = fix_keep_unknown = 1;
    fix_keep_unused = 0;
    ignore_extra = 0;
    integrity = 0;
    
    opterr = 0;
    while ((c=getopt_long(argc, argv, OPTIONS, options, 0)) != EOF) {
	switch (c) {
	case 'h':
	    fputs(help_head, stdout);
	    printf(usage, prg);
	    fputs(help, stdout);
	    exit(0);
	case 'V':
	    fputs(version_string, stdout);
	    exit(0);

	case 'b':
	    output_options &= ~WARN_BROKEN;
	    break;
	case 'c':
	    output_options |= WARN_CORRECT;
	    break;
	case 'D':
	    dbname = optarg;
	    dbext = 0;
	    break;
	case 'd':
	    output_options &= ~WARN_NO_GOOD_DUMP;
	    break;
	case 'F':
	    fix_do = 1;
	    break;
	case 'f':
	    output_options &= ~WARN_FIXABLE;
	    break;
	case 'i':
	    integrity = 1;
	    break;
	case 'K':
	    fix_keep_unknown = 1;
	    break;
	case 'k':
	    fix_keep_unknown = 0;
	    break;
	case 'L':
	    fix_keep_long = 1;
	    break;
	case 'l':
	    fix_keep_long = 0;
	    break;
	case 'n':
	    fix_do = 0;
	    fix_print = 1;
	    break;
	case 'S':
	    sample = 1;
	    break;
	case 's':
	    output_options &= ~WARN_SUPERFLUOUS;
	    break;
	case 'U':
	    fix_keep_unused = 1;
	    break;
	case 'u':
	    fix_keep_unused = 0;
	    break;
	case 'v':
	    fix_print = 1;
	    break;
	case 'w':
	    output_options &= WARN_BROKEN;
	    break;
	case 'X':
	    ignore_extra = 1;
	    break;
	case OPT_SF:
	    superfluous_only = 1;
	    break;

	default:
	    fprintf(stderr, usage, prg);
	    exit(1);
	}
    }
    
    if ((db=ddb_open(dbname, DDB_READ|dbext))==NULL) {
	myerror(ERRDB, "can't open database `%s'", dbname);
	exit(1);
    }

    if (superfluous_only) {
	if (optind != argc) {
	    fprintf(stderr, usage, prg);
	    exit(1);
	}
	
	handle_extra_files(db, dbname, sample);
	exit(0);
    }

    romhashtypes = diskhashtypes = 0;
    if (integrity) {
	/* XXX: check error */
	r_hashtypes(db, &romhashtypes, &diskhashtypes);
    }
    
    if ((nlist=r_list(db, "/list", &list)) < 0) {
	myerror(ERRDEF, "list of games not found in database `%s'", dbname);
	exit(1);
    }

    if (optind == argc) {
	for (i=0; i<nlist; i++)
	    tree_add(db, tree, list[i], sample);
    }
    else {
	for (i=optind; i<argc; i++) {
	    if (strcspn(argv[i], "*?[]{}") == strlen(argv[i])) {
		if (bsearch(argv+i, list, nlist, sizeof(char *),
			    (cmpfunc)strpcasecmp) != NULL)
		    tree_add(db, tree, argv[i], sample);
		else
		    myerror(ERRDEF, "game `%s' unknown", argv[i]);
	    }
	    else {
		found = 0;
		for (j=0; j<nlist; j++) {
		    if (fnmatch(argv[i], list[j], 0) == 0) {
			tree_add(db, tree, list[j], sample);
			found = 1;
		    }
		}
		if (!found)
		    myerror(ERRDEF, "no game matching `%s' found", argv[i]);
	    }
	}
    }

    tree_traverse(db, tree, sample);

    if (optind == argc && !ignore_extra)
	handle_extra_files(db, dbname, sample);

    return 0;
}
